### DISCUSSION TITLE
prints real time data frequency

### DISCUSSION THREAD

**Discussion by hanifhan** [grblHAL/core#307] [General]
is  grblhal can get 100Hz frequency report or it only 20Hz maximum ?

**Comment by karoria** [comment]
If you don't have specific reason to use high frequency for reports, please avoid it. I think there is no such limitation in firmware but the max workable frequency depends on which controller/hardware and means of communication (Ethernet/wifi-STA/wifi-AP/USB-serial etc.)
This also depends on content of report(e.g. 5 axis machine has large content than 3 axis as well as other selective report data)
I can use 50ms frequency (20Hz) Teensy4.1 with ethernet connection reliably but still I use 150ms duration for my machine. If you use wifi then I think 250 or 300ms should be a good choice. Remember, extreme resource utilization may make your controller lagging for other important tasks.

**Comment by terjeio** [comment]
To add to the excellent comment above: if the output buffer is full the main processing loop will be blocked until there is room for the complete report. With default transfer speed of 115200 baud in UART mode (~10Kb/s), and most drivers having a 512 byte output buffer, you can estimate how fast you can go based on the report length - which is configuration dependent. Network and native USB<sup>1</sup> connections can go (a lot) faster, but the overhead is larger so it is wise to leave plenty of headroom, especially if high step rates are configured.

Note that the auto report setting ($481) is limited to the 100-1000ms interval range.

Finally, do not forget that a sender has to process the data and that there might be a lot of "ok" messages interleaved in the output...

<sup>1</sup> Native USB is when the USB stack is handled by the MCU, not a UART <> USB bridge chip. E.g. the ESP32 driver does not use native USB.
